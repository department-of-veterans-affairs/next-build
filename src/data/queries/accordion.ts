import { QueryFormatter } from 'next-drupal-query'
import { ParagraphAccordion } from '@/types/drupal/paragraph'
import { Accordion } from '@/types/formatted/accordion'

export const formatter: QueryFormatter<ParagraphAccordion, Accordion> = (
  entity: ParagraphAccordion
) => {
  return {
    id: entity.id,
    header: entity.field_header || null,
    html: entity.field_rich_wysiwyg.processed,
  }
}
