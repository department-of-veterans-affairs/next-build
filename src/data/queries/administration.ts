import { FieldAdministration } from '@/types/drupal/field_type'

export interface Administration {
  entityId: number
  name: string | null
}

/**
 * Formats a Drupal administration field into our standardized Administration interface.
 * 
 * In our Drupal content, the `field_administration` field points to a taxonomy Term
 * that is currently called "Section", but "Section" is only a human label added to
 * it in the Drupal interface, and that name has evolved over time. It was originally,
 * "administration" (and thus the name of the field). Since the only machine name of
 * it is `field_administration`, we'll continue to call it "administration" in our code.
 */
export const formatAdministration = (
  administration: FieldAdministration | null | undefined
): Administration => ({
  entityId: administration?.drupal_internal__tid || null, // "t" is for taxonomy term
  name: administration?.name || null,
})
