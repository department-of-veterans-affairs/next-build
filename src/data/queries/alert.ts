// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { Alert } from '@/types/formatted/alert'

export const formatter: QueryFormatter<BlockAlert, Alert> = (
  entity: BlockAlert
) => {
  return {
    alertType:
      entity.field_alert_type === 'information'
        ? 'info'
        : entity.field_alert_type,
    id: entity.id,
    title: entity.field_alert_title,
    content: {
      header: entity.field_alert_content.field_text_expander,
      text: entity.field_alert_content.field_wysiwyg.processed,
    },
  }
}
