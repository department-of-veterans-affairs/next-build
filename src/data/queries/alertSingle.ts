import { QueryFormatter } from 'next-drupal-query'
import { ParagraphAlertSingle } from '@/types/drupal/paragraph'
import {
  AlertSingle,
  AlertBlock,
  AlertNonReusable,
} from '@/types/formatted/alert'
import { formatParagraph } from '@/lib/drupal/paragraphs'
import { queries } from '.'

// TODO:
//
// EDIT:
// It seems *maybe* that next-drupal-query is designed to handle this already.
// You can pass an `id` to queries.getParams().
// E.g.
// queries.getParams('paragraph--alert_single'))
//
// We don't need this here:
// export const params: QueryParams<null> = () => {
//   return queries
//     .getParams()
//     .addInclude([
//       'field_alert_block_reference',
//       'field_alert_block_reference.field_alert_content',
//       'field_alert_non_reusable_ref',
//       'field_alert_non_reusable_ref.field_va_paragraphs',
//     ])
// }
//
// ...but it seems like we might want to define
// paragraph-specific `includes` within the paragraph's
// query file (this file), and then somehow reference these in
// the node query when we need to include fields of a specific
// paragraph type.
//
// E.g.
// export const include = [
//   'field_alert_block_reference',
//   'field_alert_block_reference.field_alert_content',
//   'field_alert_non_reusable_ref',
//   'field_alert_non_reusable_ref.field_va_paragraphs',
// ]
//
// /* Some util file*/
// export const getIncludedSubFields = (
//   parentField: string,
//   subFields: string[]
// ) => [parentField, ...subFields.map((subField) => `${parentField}.${subField}`)]
//
// /* src/data/queries/resourcesSupport.ts */
// import { include } from '@/data/queries/alertSingle'
// export const params: QueryParams<null> = () => {
//   return queries.getParams().addInclude([
//     ...getIncludedSubFields('field_alert_single', include)
//   ])
// }

export const formatter: QueryFormatter<ParagraphAlertSingle, AlertSingle> = (
  entity: ParagraphAlertSingle
) => {
  return {
    type: entity.type as AlertSingle['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    alertSelection:
      entity.field_alert_selection as AlertSingle['alertSelection'],
    blockReference: queries.formatData(
      'block--alert',
      entity.field_alert_block_reference
    ) as AlertBlock,
    nonReusableRef: formatParagraph(
      entity.field_alert_non_reusable_ref
    ) as AlertNonReusable,
  }
}
