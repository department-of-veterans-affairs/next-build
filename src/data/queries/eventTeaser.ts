import { QueryFormatter, QueryParams } from 'next-drupal-query'
import { queries } from '.'
import { NodeEvent } from '@/types/drupal/node'
import { EventTeaser } from '@/types/formatted/event'
import { entityBaseFields } from '@/lib/drupal/query'

// Define the query params for fetching node--news_story.
export const params: QueryParams<null> = () => {
  return queries
    .getParams()
    .addInclude(['field_media', 'field_media.image', 'field_listing'])
}

export const formatter: QueryFormatter<NodeEvent, EventTeaser> = (
  entity: NodeEvent
) => {
  return {
    ...entityBaseFields(entity),
    date: entity.date
    // image: queries.formatData('media--image', entity.field_media), //cropType: '2_1_large'
  }
}
