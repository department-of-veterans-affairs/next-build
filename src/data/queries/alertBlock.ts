// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { AlertBlock, AlertType } from '@/types/formatted/alert'

export const formatter: QueryFormatter<BlockAlert, AlertBlock> = (
  entity: BlockAlert
) => {
  if (!entity) {
    return null
  }

  return {
    alertType: (entity.field_alert_type === 'information'
      ? 'info'
      : entity.field_alert_type) as AlertType,
    id: entity.id,
    title: entity.field_alert_title,
    content: {
      header:
        entity.field_alert_content.field_text_expander ??
        entity.field_alert_heading,
      text: entity.field_alert_content?.field_wysiwyg?.processed ?? '',
    },
  }
}
