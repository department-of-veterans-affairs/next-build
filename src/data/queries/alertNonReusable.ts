// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { AlertType, AlertNonReusable } from '@/types/formatted/alert'
import { ParagraphNonReusableAlert } from '@/types/drupal/paragraph'
import { formatParagraph } from '@/lib/drupal/paragraphs'

export const formatter: QueryFormatter<
  ParagraphNonReusableAlert,
  AlertNonReusable
> = (entity: ParagraphNonReusableAlert) => {
  if (!entity) {
    return null
  }

  return {
    type: entity.type as AlertNonReusable['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    alertType: (entity.field_alert_type === 'information'
      ? 'info'
      : entity.field_alert_type) as AlertType,
    heading: entity.field_alert_heading,
    paragraphs: entity.field_va_paragraphs.map(formatParagraph),
  }
}
