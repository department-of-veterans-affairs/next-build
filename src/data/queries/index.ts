import { createQueries } from 'next-drupal-query'
import * as NewsStory from './newsStory'
import * as NewsStoryTeaser from './newsStoryTeaser'
import * as StoryListing from './storyListing'
import * as QuestionAnswer from './questionAnswer'
import * as ExpandableText from './expandableText'
import * as LinkTeaser from './linkTeaser'
import * as MediaImage from './mediaImage'
import * as Banner from './banners'
import * as PersonProfile from './personProfile'
import * as Button from './button'
import * as AudienceTopics from './audienceTopics'
import * as Alert from './alert'
import * as EmailContact from './emailContact'
import * as BenefitsHub from './benefitsHub'
import * as Wysiwyg from './wysiwyg'
import * as StaticPathResources from './staticPathResources'
import * as HeaderFooter from './headerFooter'

export const queries = createQueries({
  // standard Drupal entity data queries
  'node--news_story': NewsStory,
  'node--news_story--teaser': NewsStoryTeaser,
  'node--story_listing': StoryListing,
  'node--q_a': QuestionAnswer,
  'node--person_profile': PersonProfile,
  'node--landing_page': BenefitsHub, // "Benefits Hub Landing Page"
  'paragraph--audience_topics': AudienceTopics,
  'paragraph--button': Button,
  'paragraph--email_contact': EmailContact,
  'paragraph--expandable_text': ExpandableText,
  'paragraph--link_teaser': LinkTeaser,
  'paragraph--rich_text_char_limit_1000': Wysiwyg,
  'paragraph--wysiwyg': Wysiwyg,
  'media--image': MediaImage,
  'block--alert': Alert,

  // Custom queries 
  'banner--alerts_lookup': Banner,
  'header-footer-data': HeaderFooter,

  // Static Path Generation
  'static-path-resources': StaticPathResources,
})
