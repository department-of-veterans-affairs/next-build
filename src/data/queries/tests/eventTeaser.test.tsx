/**
 * @jest-environment node
 */

import { NodeEvent } from '@/types/drupal/node'
import { queries } from '@/data/queries'
import mockData from '@/components/event/mock.json'
import { params } from '../eventTeaser'

const nodeEventMock: NodeEvent = mockData

describe('DrupalJsonApiParams configuration', () => {
  test('params function sets the correct include fields', () => {
    const paramsInstance = params()
    const queryString = decodeURIComponent(paramsInstance.getQueryString())
    expect(queryString).toMatch(
      /include=field_listing,field_administration,field_facility_location/
    )
  })
})

describe('event teaser formatData', () => {
  test('outputs formatted data for all fields', () => {
    const formattedData = queries.formatData(
      'node--event--teaser',
      nodeEventMock
    )
    expect(formattedData.entityId).toBe(nodeEventMock.id)
    expect(formattedData.title).toBe(nodeEventMock.title)
    expect(formattedData.fieldDescription).toBe(nodeEventMock.field_description)
  })

  test('handles missing optional fields gracefully', () => {
    const eventWithMissingFields = {
      ...nodeEventMock,
      field_address: undefined,
      field_facility_location: undefined,
    }
    const formattedDataWithMissingFields = queries.formatData(
      'node--event--teaser',
      eventWithMissingFields
    )
    expect(formattedDataWithMissingFields.fieldAddress.addressLine1).toBeNull()
    expect(formattedDataWithMissingFields.fieldFacilityLocation).toBeNull()
  })

  test('handles partial field_facility_location data correctly', () => {
    const eventWithPartialFacilityLocation: Partial<NodeEvent> = {
      ...nodeEventMock,
      field_facility_location: {
        ...nodeEventMock.field_facility_location,
        field_address: {
          langcode: 'en',
          country_code: 'US',
          administrative_area: 'VA',
          locality: 'Richmond',
          address_line1: '123 Main St',
          address_line2: 'Suite 100',
          postal_code: '23220',
        },
      },
    }

    const formattedData = queries.formatData(
      'node--event--teaser',
      eventWithPartialFacilityLocation as NodeEvent
    )
    expect(
      formattedData.fieldFacilityLocation?.entity.fieldAddress
        .administrativeArea
    ).toBe('VA')
    expect(
      formattedData.fieldFacilityLocation?.entity.fieldAddress.addressLine1
    ).toBe('123 Main St')
  })

  test('handles fully online events without a physical location', () => {
    const onlineEventMock = {
      ...nodeEventMock,
      field_location_type: 'online',
      field_url_of_an_online_event: {
        uri: '/example',
        title: 'test',
        options: [],
      },
      field_facility_location: undefined,
    }
    const formattedData = queries.formatData(
      'node--event--teaser',
      onlineEventMock
    )
    expect(formattedData.fieldUrlOfAnOnlineEvent).toBe(
      onlineEventMock.field_url_of_an_online_event
    )
    expect(formattedData.fieldFacilityLocation).toBeNull()
  })
})
