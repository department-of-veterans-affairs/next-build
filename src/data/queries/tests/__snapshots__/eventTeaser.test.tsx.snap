// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event teaser formatData handles event with full fields, including address and facility location 1`] = `
{
  "changed": "2023-08-16T15:20:30+00:00",
  "entityBundle": "node--event",
  "entityId": "16349f16-be65-46e3-9660-1d3d598a4a0b",
  "entityPublished": true,
  "entityUrl": {
    "path": "/central-iowa-health-care/events/52265",
  },
  "fieldAdditionalInformationAbo": null,
  "fieldAdditionalListings": null,
  "fieldAddress": {
    "addressLine1": "9000 Douglas Ave",
    "addressLine2": null,
    "administrativeArea": "IA",
    "countryCode": "US",
    "locality": "Urbandale",
    "postalCode": null,
  },
  "fieldAdministration": {
    "entity": {
      "entityId": "3a530a55-e922-48de-b18c-9c9f92a9ae98",
    },
  },
  "fieldBody": {
    "format": "rich_text",
    "processed": "<p>Pickleball Club</p>

<p>Meets Thursdays from 9 to 11 a.m. </p>

<p>Contact Kay Queck (515) 214-4578</p>",
    "value": "<p>Pickleball Club</p>

<p>Meets Thursdays from 9 to 11 a.m.&nbsp;</p>

<p>Contact Kay Queck (515) 214-4578</p>
",
  },
  "fieldCtaEmail": "test@va.gov",
  "fieldDatetimeRangeTimezone": [
    {
      "duration": 120,
      "endTime": "2023-09-07T16:00:00+00:00",
      "endValue": 1694102400,
      "end_value": "2023-09-07T16:00:00+00:00",
      "rrule": 180,
      "rrule_index": 1,
      "startTime": "2023-09-07T14:00:00+00:00",
      "timezone": "America/Chicago",
      "value": 1694095200,
    },
    {
      "duration": 120,
      "endTime": "2023-09-14T16:00:00+00:00",
      "endValue": 1694707200,
      "end_value": "2023-09-14T16:00:00+00:00",
      "rrule": 180,
      "rrule_index": 2,
      "startTime": "2023-09-14T14:00:00+00:00",
      "timezone": "America/Chicago",
      "value": 1694700000,
    },
  ],
  "fieldDescription": "Pickleball ",
  "fieldEventCost": "Free",
  "fieldEventCta": null,
  "fieldEventRegistrationrequired": false,
  "fieldFacilityLocation": null,
  "fieldFeatured": false,
  "fieldHowToSignUp": "email",
  "fieldLink": null,
  "fieldListing": {
    "entity": {
      "entityId": "55cb823c-df6d-4f77-814d-6024e08cd4f6",
    },
  },
  "fieldLocationHumanreadable": "Walker Johnston Park",
  "fieldLocationType": "non_facility",
  "fieldOrder": null,
  "fieldUrlOfAnOnlineEvent": null,
  "title": "Pickleball Club",
}
`;

exports[`event teaser formatData outputs formatted data 1`] = `
{
  "changed": "2023-08-16T15:20:30+00:00",
  "entityBundle": "node--event",
  "entityId": "16349f16-be65-46e3-9660-1d3d598a4a0b",
  "entityPublished": true,
  "entityUrl": {
    "path": "/central-iowa-health-care/events/52265",
  },
  "fieldAdditionalInformationAbo": null,
  "fieldAdditionalListings": null,
  "fieldAddress": {
    "addressLine1": "9000 Douglas Ave",
    "addressLine2": null,
    "administrativeArea": "IA",
    "countryCode": "US",
    "locality": "Urbandale",
    "postalCode": null,
  },
  "fieldAdministration": {
    "entity": {
      "entityId": "3a530a55-e922-48de-b18c-9c9f92a9ae98",
    },
  },
  "fieldBody": {
    "format": "rich_text",
    "processed": "<p>Pickleball Club</p>

<p>Meets Thursdays from 9 to 11 a.m. </p>

<p>Contact Kay Queck (515) 214-4578</p>",
    "value": "<p>Pickleball Club</p>

<p>Meets Thursdays from 9 to 11 a.m.&nbsp;</p>

<p>Contact Kay Queck (515) 214-4578</p>
",
  },
  "fieldCtaEmail": "test@va.gov",
  "fieldDatetimeRangeTimezone": [
    {
      "duration": 120,
      "endTime": "2023-09-07T16:00:00+00:00",
      "endValue": 1694102400,
      "end_value": "2023-09-07T16:00:00+00:00",
      "rrule": 180,
      "rrule_index": 1,
      "startTime": "2023-09-07T14:00:00+00:00",
      "timezone": "America/Chicago",
      "value": 1694095200,
    },
    {
      "duration": 120,
      "endTime": "2023-09-14T16:00:00+00:00",
      "endValue": 1694707200,
      "end_value": "2023-09-14T16:00:00+00:00",
      "rrule": 180,
      "rrule_index": 2,
      "startTime": "2023-09-14T14:00:00+00:00",
      "timezone": "America/Chicago",
      "value": 1694700000,
    },
  ],
  "fieldDescription": "Pickleball ",
  "fieldEventCost": "Free",
  "fieldEventCta": null,
  "fieldEventRegistrationrequired": false,
  "fieldFacilityLocation": null,
  "fieldFeatured": false,
  "fieldHowToSignUp": "email",
  "fieldLink": null,
  "fieldListing": {
    "entity": {
      "entityId": "55cb823c-df6d-4f77-814d-6024e08cd4f6",
    },
  },
  "fieldLocationHumanreadable": "Walker Johnston Park",
  "fieldLocationType": "non_facility",
  "fieldOrder": null,
  "fieldUrlOfAnOnlineEvent": null,
  "title": "Pickleball Club",
}
`;
